File: saul-examples/src/main/java/edu/illinois/cs/cogcomp/saulexamples/setcover/SetCoverSolver.java
package edu.illinois.cs.cogcomp.saulexamples.setcover;


public class SetCoverSolver {
    public static void main(String[] args) {

        containsStationConstrained classifier = new containsStationConstrained();

        for (String file : args) {
            City c = new City(file);
            for (Neighborhood n : c.getNeighborhoods()) {
                System.out.println(n.getNumber() + ": " + classifier.classify(n));
            }
        }
    }
}

File: saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/setcover/SetCoverSolver.scala
package edu.illinois.cs.cogcomp.saulexamples.setCover

import edu.illinois.cs.cogcomp.saul.classifier.ConstrainedClassifier
import edu.illinois.cs.cogcomp.saul.constraint.ConstraintTypeConversion._
import edu.illinois.cs.cogcomp.saul.datamodel.DataModel

import scala.collection.mutable.{ Map => MutableMap }

class SetCoverSolverDataModel extends DataModel {

  val cities = node[City]

  val neighborhoods = node[Neighborhood]

  val cityContainsNeighborhoods = edge(cities, neighborhoods, 'cityID)

  cityContainsNeighborhoods.populateWith((c, n) => c == n.getParentCity)
}

object containsStationConstraint extends ConstrainedClassifier[Neighborhood, City](setCoverApp.trainingData, new ContainsStation()) {

  override def subjectTo = setCoverApp.containsStationConstrint

  //    ConstraintClassifier.constraintOf[City]({
  //    x: City => {
  //      val containStation = new ContainsStation()
  //      x.getNeighborhoods _forAll {
  //        n: Neighborhood => {
  //          (containStation on n isTrue) ||| (
  //            n.getNeighbors _exists {
  //              n2: Neighborhood => containStation on n2 isTrue
  //            })
  //        }
  //      }
  //    }
  //  })
}
